agent/cache-types/streaming_health_services.go

```diff
@@ -98,6 +109,64 @@ func (c *StreamingHealthServices) Fetch(opts cache.FetchOptions, req cache.Reque
 	return state.Fetch(opts)
 }
 
+// This doesn't correspond to a streaming topic, but rather it looks at two
+// other topics and synthesizes results from them.
+func (c *StreamingHealthServices) preferConnectFetch(origReq *structs.ServiceSpecificRequest) (cache.FetchResult, error) {
+	// First we prefer returning connect enabled endpoints for this service:
+	var maxIndex uint64
+	{
+		connectReq := *origReq
+		connectReq.PreferConnect = false
+		connectReq.Connect = true
+
+		connectIdx, connectResults, err := c.fetchFromCache(&connectReq)
+		if err != nil {
+			return cache.FetchResult{}, err
+		}
+
+		maxIndex = connectIdx
+
+		if len(connectResults.Nodes) > 0 {
+			var result cache.FetchResult
+			result.Value = connectResults
+			result.Index = connectIdx
+			return result, nil
+		}
+	}
+
+	plainReq := *origReq
+	plainReq.PreferConnect = false
+	plainReq.Connect = false
+
+	plainIdx, plainResults, err := c.fetchFromCache(&plainReq)
+	if err != nil {
+		return cache.FetchResult{}, err
+	}
+
+	if plainIdx > maxIndex {
+		maxIndex = plainIdx
+	}
+
+	var result cache.FetchResult
+	result.Value = plainResults
+	result.Index = maxIndex
+	return result, nil
+}
+
+func (c *StreamingHealthServices) fetchFromCache(req *structs.ServiceSpecificRequest) (uint64, *structs.IndexedCheckServiceNodes, error) {
+	// Background is fine here because this isn't a blocking query as no index is set.
+	// Therefore this will just either be a cache hit or return once the non-blocking query returns.
+	rawResult, rm, err := c.cache.Get(context.Background(), StreamingHealthServicesName, req)
```

Comment:
I cribbed this from how the ConnectCALeaf cache-type grabs the roots. I suspect there might be something fishy about this pattern used here.